core/src/main/scala/it/scalachess/core/logic/moves/generators/package.scala

```diff
@@ -0,0 +1,63 @@
+package it.scalachess.core.logic.moves
+
+import it.scalachess.core.Color
+import it.scalachess.core.board.{ Board, Position }
+import it.scalachess.core.pieces.PieceType
+import scalaz.{ Failure, Success, Validation }
+
+import scala.annotation.tailrec
+
+package object generators {
+
+  def generateSimpleMove(pieceType: PieceType,
+                         player: Color,
+                         board: Board,
+                         from: Position,
+                         to: Option[Position]): Validation[String, ValidSimpleMove] =
+    to match {
+      case None => Failure(s"${pieceType.name}'s movement: the end position doesn't exist in the board")
+      case Some(to) =>
+        board.pieceAtPosition(to) match {
+          case None => Success(ValidSimpleMove(from, to, pieceType, player, None))
+          case Some(pieceToCapture) =>
+            pieceToCapture.color match {
+              case player.other => Success(ValidSimpleMove(from, to, pieceType, player, Some(to)))
+              case _            => Failure(s"${pieceType.name}'s movement: can't attack an ally piece")
+            }
+        }
+    }
+
+  @tailrec
+  def generateLinearMovementSimpleMoves(pieceType: PieceType,
```

Comment:
this method should be private and the public one should call this providing an empty List to begin the recursion